Add Kotlin support, with minimal intrusion in other's files

Lakedaemon opened the pull request:
This will hopefully work with Travis...

Lakedaemon commented:
With packages fixed, kotlin support now works properly. 

Next, I should improve the performance of Enums with at least : 
+ a function for Unions or values in an arithmetric progression
+ a switch for say less than 8 enums
+ a map or a binary search otherwise ?

I would also like for travis to launch and check the kotlin tests but don't know how to do that (or if I even can)

Lakedaemon commented:
Improved the performance of enums when they are in an arithmetic progression (function) and with a map. I included some new enums in the monster schema to build the new implementations but haven't writtent tests for them yet. :/

By the way, why does flatbuffers serialize/deserialize enums with their value instead of their ordinal ?

Lakedaemon commented:
A change to accomodate the fact that the recently released Kotlin 1.0 Beta 3 undeprecated values() and deprecated values : https://github.com/JetBrains/kotlin/releases

Hopefully Kotlin 1.0 will be released soon and quit being a moving target (it's not that difficult to hit, but still...) 

Also, structRaw now returns an offset to the struct (so that you can create a flatbuffer that is a struct conveniently)

gwvo commented:
It's starting to shape up. Still not super happy about the fact that this is not part of idl_general.cpp, but I can see the language is different enough where this would not be easy. Opinions anyone? @rw ?

Do you know anyone in the Kotlin community that is somewhat of an expert on the language that would want to help with a code review of this PR?

Lakedaemon commented:
We can just ask the people at jetbrain to look at it. They are really helpfull and are usually really happy whenever a new library pop up for Kotlin. 

Also, though we all want to avoid code duplication and maintenance hell. I still don't see how the solution involving idl_general.cpp and ifs (spagetty code instead of overloading) enables that : 
especially if there are going to be code generators for java 1.7, java 1.8, java 1.9, java 1.10, kotlin 1.0, kotlin 2.0...

Having cleanly separate files for each language makes it so much more easier to implement features/fix things

Lakedaemon commented:
I found and fixed a few issues with my code related to enums and to downsizing constants... 
(-2.toShort() produces an error in Kotlin when (-2).toShort() doesn't, which feels slightly weird to me)
I added a few fields to the Monster Table at the end.

Next, I should test mutations of these fields, implement enums with a sparse array like in C++ (and depending on the range of enum values use that, binarySearch or a map).
Maybee we should have all types of vectors/fields in the Monster table to test all generated methods...

Lakedaemon commented:
Finally, I wrote a base class for a Code Generator and rebased the Kotlin Code Generator on it.
I used the opportunity to clean the code wherever I could. But there are still improvements to do.

I would like to get feedback about it

Next, I'll make the base class on a diet.
 
It will probably be very general, with just :
a public constructor,  
a public method with signature bool generate() 
6 const protected members (parser, path, file_name, keywords, namespace_components, namespace_dir)

On top of that, we might write (for example) a StronglyTypedCodeGenerator class/trait 
with more utility functions like the safeName method that avoids name clash for fields
or a safeType method that would avoid name clash for types...

And then the various Code Generators for various language
like Kotlin1, Kotlin 1.2...

Lakedaemon commented:
Split the former base class into the most simple/general class for code generation. Namely
  class BaseGenerator {
    public:
      BaseGenerator(const Parser & parser_,const std::string & path_,const std::string & file_name_) : parser(parser_), path(path_), file_name(file_name_) {};
      virtual bool generate() = 0;
    protected:
      virtual ~BaseGenerator() {};
      
      const Parser & parser;
      const std::string & path;
      const std::string & file_name;
  };

All code generators can migrate very easily to this new class (this can be done in less than an hour, without hassle and the intervention of any code generator maintainer)

Though it is simple, there are some immediate benefits : 
1) No need to pass those parser, path, file_name arguments in all functions.
2) Easier for code generator writers to start writing a code generator for their exotic language : 
they aren't  necessary fluent in C++, it took me a lot of time/frustration at first to understand how to make those paranoid errors like "parameter isn't used" disappear...
3) Standardisation (nice for cross pollenisation) : now all code generators will start looking the same (which is good), with the important variables named the same way...
4) Inheritance is needed to be able to share code between code generators for different versions of the same language 

In my opinion, In the end, it will make developing, maintaining and sharing code easier
Also, it might  help in one of your goal : "somehow, make it so that a feature of flatbuffers can be implemented simply in parallel in a lot of languages"

This might come with the help of other orthogonal classes/interfaces

Lakedaemon commented:
Kotlin 1.0.0 has been released today (much sooner than I expected).

I'll make the modification today to use arrays instead of varargs, make a last push an ask some of the jet brain people for a review...they'll get delighted for another library to get native kotlin support on day 1

Lakedaemon commented:
I also simplified some method's names
names for methods aren't final. I would like them to be as simple and explicit as possible. 
Also, I'm waiting for the feedback of others to rename them

I'll be asking some of the jetbrains people to look at this pr now, for a review

Lakedaemon commented:
Now that the latest code pass the Travis check, I asked in a tweet the help of a kotlin expert to help review this pull request. Let's just hope now that I have made a good job at it. ^^

Lakedaemon commented:
today, because I wasn't using named parameters in the constructor, I got burned by this simple schema : 

namespace org.lakedaemon.setup;

table LongsCache {
    size:int; // file size
    blueprint : string (required); // tells if the cache is valid
    longs:[long] (required);
}

root_type LongsCache;

This happens because offsets are standard ints in the kotlin generator.
-> I think that I'll start writing a kotlin generator that uses Generics (like the csharp generator, to differenciates between offsets for table, vectors and standard ints).
This will be a good experiment to see if I can reuse code from the kotlin 1.0 generator

It will allow people to check their generated code (by swaping the different kotlin generators, the safe/slow one and the less safe/fast one)
Also, when/if kotlin implements type aliasing, it might bring improved safety without performance penalty

In my next commit, I'll add the generated kotlin classes (I see now the wisdom of that)

Sadly, I have had no news from the Jet brain people yet, for a review of this pr.

gwvo commented:
We don't use generics in Java, because wrapping an int offset in an Object is very expensive, not a cost we want to pay. C# has by-value structs which makes this kind of wrapping cheap. If Kotlin implements generics like Java, I'd advice against using them.

Lakedaemon commented:
Kotlin implements Generics mostly like java (except that they behave way better when inlined), this is why I used Int for offsets in the kotlin 1.0.0 generator.

In the generator for kotlin 1.0.0, The create functions for Tables return a lambda (with the signature FlatbufferBuilder.() -> Int) instead of an Int because they can be inlined (no performance penalty) and allow a very nice way to create a Table in one line

If/when the people at Jet Brains implement type alias for the next version of Kotlin (they wrote that they were interested in doing that), it might be possible to make those create function return a type alias instead (like Offset<YourTable>) for this lambda bringing a lot more type safety (again without performance penalty/without object creation) 

At least, this is my hope. It remains to be seen if it can be done.
Maybee I should wait before experimenting with that...

Lakedaemon commented:
Saddly, I haven't been able to peek at the work of @evolutional yet. So, I gave a try at writing a CodeWriter class and refactored the kotlin code generator so that it uses it.

It works and produces code that is more beautiful land easier to maintain. 
It only supports indentation for now, I'll give a go at supported C++ like namespaces later.

I'm partially satisfied with the use of operator += and +
I'll probably have to switch to an operator that is left-assossiative like << 
because code << aaa << bbb is better than 
code = code + aaa + bbb

Once I do this switch and I am satisfied, I'm going to look at splitting and refactoring the java code generator

gwvo commented:
First, lets keep the task of writing a Kotlin implementation separate from other tasks, such as factoring out general code generator functions.

I don't think we need something like CodeWriter, it just wraps the current string concatenation for no obvious gain. Lets keep things simple with a very minimal, incremental transition.

Also, the formatting and indentation of the C++ code is all over the place. Maybe run clang-format (with the Google style) on it?

Lakedaemon commented:
concatenation for std::string won't give you semi-automatic indentation, semi-automatic namespace, dispatch code to different files (say 1 file for a class, 1 file for top level methods...)
With std::string, I had to put \n and \t everywhere and the code generated still wasn't as nice as the last batch.

I'm experimenting in my branch to see what api would be best for code generator maintainers.
I don't really expect anymore this pull request to be merged as it, indeed, does too many things.
It is nice to get some feedback though.

If I somehow manage to produce a nice simple way to factor out general code generators and if you are interested to merge it in the flatbuffers codebase, it'll still be possible to split it in small, manageable pull requests

In the mean time, I have to experiment and try to "port" another generator to see if the ideas behind the code are sound (also eating my own dog food is a good way to see if I'm doing something wrong or not and to find issues early)

I didn't know about clang-format. I will look into it : I use no ide for C++, just jedit and syntax highlighting. Manually maintaining indentation is really tedious, so I gave up

Lakedaemon commented:
I setup clang-format-3.7 (on ubuntu)
tried : clang-format-3.7 -style=Google -i ./idl_gen_kotlin.cpp

And it made things worse... (good bye indentations....)

I tried again after : clang-format-3.7 -style=google -dump-config > .clang-format
No improvements

gwvo commented:
That may well be, but we have a lot of generators, and we want a transition path that is as incremental as possible. Your class is going to be a very intrusive change.

What did it make worse? It enforces the Google style guide, which not everyone likes, but this project adheres to.

Lakedaemon commented:
You are right. I guess that I am among the people that don't like the way code looks with the Google style guide. 

My future prs will follow the Google style guide as this project adheres to it (I knew it before but I didn't know about clang-format). 

Also, I will make it so (or die trying) that the transition path is as smooth/simple and unobstrusive as possible. 

I have gained quite a lot of experience transitioning my code to this CodeWriter and I can use that to make it simple to code generators maintainers.

As of now, transitioning mostly means : 
1) removing lots of "\n" "\t" in strings and replacing them by NL (newline), TAB (tab to the right) and BAT (pop the last tab amount). I can change the names if you don't think they are the best for the job) 

The python guys are already doing it this way I think

2) in some methods signature, replacing std::string with CodeWriter

3) instancing a CodeWriter somewhere and saving to file at the end of the process

Later on, the C++ guys will gain better namespace formatting, for free.

I think that I'll let the += and + operators overrides in CodeWriter 
(nice for transition) 
but I'll probably add a << override to append stuff (and command tabs/namespaces) to the codeWriter

The problem with += and + is that you have unexpected results if you do something like 
code = a + b + c  instead of code = code a+b +c 
code += a + b instead of code += a; code += b 
code + a instead of code += a;

You don't have those problems with << 

Also, I know that I am annoying ^^
Please bear with me. I'll soon propose in a new PR branch some small PR to start the transitioning process and you shall judge (and probably ask changes).

gwvo commented:
I suggested not using CodeWriter at all at this point. I would focus on just getting Kotlin right, and leave any general code generator refactorings till later commits. Or, suggest a refactoring for the existing generators as a PR that is completely independent from Kotlin, if you prefer.

Lakedaemon commented:
Well, then I will work under your supervision to get the Kotlin code generator right in a git branch, 
without the CodeWriter (I'll still develop and use it for my own personal projects so, if you ever warm to it, it'll be available).

Could you guide me on what needs to be done to get it right/merged ? 
(starting from commit https://github.com/google/flatbuffers/pull/1190/commits/40ad6f195986b81d1b022ac7629e9a08093b119b)

Feature wise, the Kotlin code generator should be on par with the Java code generator (compare the generated classes ?) or slightly better (toString method, better enums/unions, better api...). 
I have been using  in my own projects for a few months and I'm happy with it. 
I would really welcome some feedback from other users

gwvo commented:
"starting from commit 40ad6f1".. can you make the PR exactly what you're trying to merge, so it is easier for me to review?

Again, not an expert in Kotlin, so will only comment on general structure. Would like to have the ok from a more expert Kotlin user before merging.

Lakedaemon commented:
I found a way to make the kotlin api even safer with the createStruct/Tables/Vectors methods.

For offset, instead of asking for an Int (error prone, not safe) or for a lambda ()->Int (nice for generating structs inline but still error prone) I am now asking for a lambda (Struct?)->Int.

So, the compiler will now make sure that the right offsets are used in the right places.
Also, there is still no performance penalty because the lambda is inlined and it is really simple.

I have already implemented that for tables and structs. 
It will take me a little more time to implement it for vectors/unions. 
Once it is done, I'll try to make a clean pr (google style) without the CodeWriter. And I'll ask the JetBrains people again for a review.

gwvo commented:
I would guess a lambda is not inlined if it is passed around by value.

Safety is a very worthy goal. In the case of FlatBuffers, efficiency trumps it though.

Lakedaemon commented:
On 04/19/2016 12:52 AM, Wouter van Oortmerssen wrote:
>
> I would guess a lambda is not inlined if it is passed around by value.
>

Yes indeed. This is something I'm thinking about (especially for String).

val mon = monster(of("MyMonster"), // <-------- i'm considering making 
string offsets safe, not sure yet.
                     pos = test(2.toShort(), 3.toByte()) // <---- this 
struct is safer with a lambda that is inlined
)


val pos = test(2.toShort(), 3.toByte()) // <---- there is object 
creation there

The object constructed is very light though as the lambda is just {offset}


val mon = monster(of("MyMonster"), // <-------- i'm considering making 
string offsets safe
                     pos = pos, // <--- the lambda is inlined there
)



It's kind of hard to get the right balance between safety, 
user-friendliness and efficiency

I would guess that building a Flatbuffer is quite an heavy operation and 
I'm not sure if the cost of creating a very thin wrapper around an Int : 
the JVM should be quite heavily optimize use and garbage collection of 
very short objects
(also kotlin doesn't have int (primitive type), it only has Int (object 
type)) is really detrimental to performances.
We would need the advice of a Kotlin expert for this/profiling.


> Safety is a very worthy goal. In the case of FlatBuffers, efficiency 
> trumps it though.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/google/flatbuffers/pull/1190#issuecomment-211618074>
>

aardappel commented:
Closing due to inactivity. Feel free to re-open.
